File: py_impl/meatloaf.py
from raw.bread.errors import MeatloafError
import time

from raw.bread.interpreter import Interpreter, NamespaceTable, Stack
from raw.bread.parser import Parser
from raw.bread.lexer import Lexer
from time import time

"""
TODO: Make Interpreted errors better
"""
print("\nWelcome to the Meatloaf interactive grill\n Get ready to make your meat and we'll bread em.\nVersion 0.0.0.a\n")
global_namespace = NamespaceTable()
global_stack = Stack("<program>", namespace=global_namespace)

while True:
    start = time()

    status = Lexer.tokenize(input(f"->- "))
    if isinstance(status, MeatloafError):
        print("...", repr(ast).replace("\n", "\n    ", -1))
        continue
    else:
        print("...", status)

    ast = Parser.parse(status)
    if isinstance(ast, MeatloafError):
        print("...", repr(ast).replace("\n", "\n    ", -1))
        continue
    else:
        print("...", ast)

    results = Interpreter.interpret(global_stack, ast)
    print(results)

File: src/Meatloaf.cpp
#include <stdio.h>
#include <iostream>
#include <sstream>
#include <vector>
#include <memory>

#include "Lexer.h"
#include "Parser.h"

int main()
{
    std::cout << "Welcome to the Meatloaf interactive grill\n Get ready to make your meat and we'll bread em.\nVersion 0.0.0.a\n"
              << std::endl;
    std::string cmdlet;
    while (true)
    {
        std::cout << "->- ";
        std::getline(std::cin, cmdlet);

        std::vector<Token> tokens = Lexer::tokenize(cmdlet);
        std::cout << "... ";
        for (Token tk : tokens)
        {
            std::cout << tk.to_string() << ' ';
        };
        std::cout << tokens.size() << '\n';
        Parser p{tokens};
        std::vector<Operable*> nodes = p.exact();
        std::cout << "... " << nodes.size() << ' ';
        std::cout << nodes[0]->to_string() << ' ';
        std::cout << std::endl;
    };
    std::cout << std::endl;
    return 0;
};
